srml/support/src/storage/generator.rs

```diff
@@ -194,6 +217,15 @@ pub trait StorageMap<K: codec::Codec, V: codec::Codec> {
 	fn mutate<R, F: FnOnce(&mut Self::Query) -> R, S: Storage>(key: &K, f: F, storage: &S) -> R;
 }
 
+/// A `StorageMap` with enumerable entries.
+pub trait EnumerableStorageMap<K: codec::Codec, V: codec::Codec>: StorageMap<K, V> {
+	/// Return current head element.
+	fn head<S: Storage>(storage: &S) -> Option<K>;
+
+	/// Enumerate all elements in the map.
+	fn enumerate<'a, S: Storage>(storage: &'a S) -> Box<Iterator<Item = (K, V)> + 'a>;
```

Comment:
Why do we want to have a `Box` here? An associated type work as well and would not require `Box` :)

Reply:
That was my first approach, but you can't have lifetimes in the associated types (and we require it because of  `&Storage`)

Also `impl Iterator<> + 'a` is not an option, @tomaka would you like to whine a bit on this one :)